edb/server/compiler/compiler.py

```diff
@@ -2012,6 +2008,14 @@ def _try_compile(
                 raise errors.InternalServerError(
                     f'QueryUnit {unit!r} is cacheable but has config/aliases')
 
+            multi_card = unit.cardinality in (
+                enums.Cardinality.MANY, enums.Cardinality.AT_LEAST_ONE,
+            )
+            if multi_card and ctx.expected_cardinality_one:
```

Comment:
Took me a while to figure out that I need to look at: https://github.com/edgedb/edgedb/blob/ad92c48d26f9052371dfc48d31e414f592e16485/edb/server/compiler/compiler.py#L1854-L1855

so that non-trailing query units have the cardinality `NO_RESULT`. Would be more readable if we can just check `rv[-1]`, or add a comment?

Reply:
Oh, yeah, I actually meant to be doing this logic on `rv` itself!